File: pruebas/generar_pptx.py
"""
Genera la presentacion PowerPoint de pruebas PremiumBus.
Requiere: pip install python-pptx
Ejecutar: python generar_pptx.py
"""
from pptx import Presentation
from pptx.util import Inches, Pt, Emu
from pptx.dml.color import RGBColor
from pptx.enum.text import PP_ALIGN, MSO_ANCHOR
import os

prs = Presentation()
prs.slide_width = Inches(13.333)
prs.slide_height = Inches(7.5)

DARK_BG = RGBColor(0x0F, 0x17, 0x2A)
BLUE = RGBColor(0x1A, 0x3A, 0x6B)
BLUE2 = RGBColor(0x2B, 0x5E, 0xA7)
GREEN = RGBColor(0x22, 0xC5, 0x5E)
RED = RGBColor(0xEF, 0x44, 0x44)
WHITE = RGBColor(0xFF, 0xFF, 0xFF)
GRAY = RGBColor(0x94, 0xA3, 0xB8)
LGRAY = RGBColor(0xE2, 0xE8, 0xF0)
CARD = RGBColor(0x1E, 0x29, 0x3B)

def set_bg(slide, color=DARK_BG):
    bg = slide.background
    fill = bg.fill
    fill.solid()
    fill.fore_color.rgb = color

def add_text(slide, text, x, y, w, h, size=12, color=LGRAY, bold=False, align=PP_ALIGN.LEFT):
    txBox = slide.shapes.add_textbox(Inches(x), Inches(y), Inches(w), Inches(h))
    tf = txBox.text_frame
    tf.word_wrap = True
    p = tf.paragraphs[0]
    p.text = text
    p.font.size = Pt(size)
    p.font.color.rgb = color
    p.font.bold = bold
    p.alignment = align
    # Handle newlines
    if '\n' in text:
        tf.clear()
        for i, line in enumerate(text.split('\n')):
            if i == 0:
                p = tf.paragraphs[0]
            else:
                p = tf.add_paragraph()
            p.text = line
            p.font.size = Pt(size)
            p.font.color.rgb = color
            p.font.bold = bold
            p.alignment = align

def add_card(slide, x, y, w, h):
    shape = slide.shapes.add_shape(
        5, Inches(x), Inches(y), Inches(w), Inches(h)
    )
    shape.fill.solid()
    shape.fill.fore_color.rgb = CARD
    shape.line.fill.background()

# ===== SLIDE 1: Portada =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s, BLUE)
add_text(s, 'Plan de Pruebas de Software', 0.5, 1.5, 12.3, 1, 40, WHITE, True, PP_ALIGN.CENTER)
add_text(s, 'PremiumBus — Metodologia Pressman (Cap. 17-20)', 0.5, 2.8, 12.3, 0.5, 20, LGRAY, False, PP_ALIGN.CENTER)
add_text(s, '10 pruebas ejecutadas  |  10 aprobadas  |  100% cobertura', 0.5, 3.6, 12.3, 0.5, 14, GRAY, False, PP_ALIGN.CENTER)
add_text(s, 'Mayo 2026', 0.5, 5.5, 12.3, 0.3, 13, GRAY, False, PP_ALIGN.CENTER)

# ===== SLIDE 2: Autenticacion =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s)
add_text(s, 'Modulo de Autenticacion', 0.5, 0.2, 12, 0.6, 28, WHITE, True)
add_card(s, 0.4, 1, 6, 5.8)
add_text(s, 'CP-AUTH-01: Login y Retencion de Token', 0.6, 1.1, 5.5, 0.4, 14, BLUE2, True)
add_text(s, 'Se ingresaron credenciales validas.\nSe verifico acceso al Home.\nSe simulo cierre y reapertura de la app.\n\nEl token se mantiene en localStorage.\nLa sesion persiste sin pedir credenciales.\n\nResultado: APROBADA', 0.6, 1.6, 5.5, 4.5, 12, LGRAY)
add_card(s, 6.8, 1, 6, 5.8)
add_text(s, 'CP-AUTH-02: Prevencion XSS/SQL', 7, 1.1, 5.5, 0.4, 14, BLUE2, True)
add_text(s, 'Se inyectaron 3 vectores de ataque:\n- <script>alert(1)</script>\n- \' OR 1=1 --\n- <img onerror=alert(1)>\n\nTodos bloqueados por validacion regex\nantes de enviar peticion de red.\n(Early Return Pattern)\n\nResultado: APROBADA', 7, 1.6, 5.5, 4.5, 12, LGRAY)

# ===== SLIDE 3: Compras =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s)
add_text(s, 'Modulo de Compras — Valores Limite', 0.5, 0.2, 12, 0.6, 28, WHITE, True)
add_card(s, 0.4, 1, 6, 5.8)
add_text(s, 'CP-COMPRA-01: Saldo Exacto $13.50', 0.6, 1.1, 5.5, 0.4, 14, GREEN, True)
add_text(s, 'Saldo del usuario = $13.50\nPrecio del boleto = $13.50\n\nSe selecciono ruta, asiento y se confirmo.\nLa transaccion fue aprobada.\nCodigo QR emitido.\n\nSaldo final: $0.00 exacto.\n\nResultado: APROBADA', 0.6, 1.6, 5.5, 4.5, 12, LGRAY)
add_card(s, 6.8, 1, 6, 5.8)
add_text(s, 'CP-COMPRA-02: Saldo Insuficiente $13.49', 7, 1.1, 5.5, 0.4, 14, RED, True)
add_text(s, 'Saldo = $13.49 (1 centavo bajo el limite)\nPrecio = $13.50\nDiferencia = -$0.01\n\nToast rojo: "Fondo insuficiente"\nTransaccion abortada.\n\nSaldo preservado sin deduccion.\n\nResultado: APROBADA', 7, 1.6, 5.5, 4.5, 12, LGRAY)

# ===== SLIDE 4: Roles =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s)
add_text(s, 'Modulo de Roles y Acceso', 0.5, 0.2, 12, 0.6, 28, WHITE, True)
add_card(s, 0.4, 1, 6, 5.8)
add_text(s, 'CP-ROL-01: Restriccion Vistas Admin', 0.6, 1.1, 5.5, 0.4, 14, BLUE2, True)
add_text(s, 'Con cuenta Admin logueada se verifico:\n\nPanel Conductores -> VISIBLE\nEstadisticas -> VISIBLE\nBoton Comprar -> OCULTO\nQuick Action Comprar -> OCULTO\n\nRenderizado condicional correcto.\n\nResultado: APROBADA', 0.6, 1.6, 5.5, 4.5, 12, LGRAY)
add_card(s, 6.8, 1, 6, 5.8)
add_text(s, 'CP-ROL-02: Directorio de Usuarios', 7, 1.1, 5.5, 0.4, 14, BLUE2, True)
add_text(s, 'Se navego al directorio como Admin.\n\nLista completa cargada con\nnombre, correo y rol de cada usuario.\n\nSin errores de composite indices\nen consola de Firebase.\n\nResultado: APROBADA', 7, 1.6, 5.5, 4.5, 12, LGRAY)

# ===== SLIDE 5: Perfil =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s)
add_text(s, 'Modulo de Perfil y Rendimiento', 0.5, 0.2, 12, 0.6, 28, WHITE, True)
add_card(s, 0.4, 1, 12.4, 5.8)
add_text(s, 'CP-PERF-01: Carga de Fotografia de Perfil', 0.6, 1.1, 12, 0.4, 14, BLUE2, True)
add_text(s, 'Se abrio el selector de imagen.\nSe cargo foto JPG valida (1.2 MB).\n\nDurante la subida se mostro barra de progreso\nasincrona (0% a 100%).\n\nAl completar, la miniatura del avatar se actualizo\nen tiempo real sin recargar la pantalla.\n\nIntegracion con Firebase Storage validada.\n\nResultado: APROBADA', 0.6, 1.6, 12, 4.5, 13, LGRAY)

# ===== SLIDE 6: Geolocalizacion =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s)
add_text(s, 'Modulo de Geolocalizacion', 0.5, 0.2, 12, 0.6, 28, WHITE, True)
add_card(s, 0.4, 1, 6, 5.8)
add_text(s, 'CP-GEO-01: Tracking en Tiempo Real', 0.6, 1.1, 5.5, 0.4, 14, BLUE2, True)
add_text(s, 'Con GPS activado se abrio el mapa.\n\nEl marcador del bus se desplazo\nfluidamente sin saltos.\n\nCoordenadas y velocidad actualizadas\nen tiempo real.\n\nPoller optimizado.\n\nResultado: APROBADA', 0.6, 1.6, 5.5, 4.5, 12, LGRAY)
add_card(s, 6.8, 1, 6, 5.8)
add_text(s, 'CP-GEO-02: Corte de GPS', 7, 1.1, 5.5, 0.4, 14, RED, True)
add_text(s, 'Se desactivo GPS durante tracking.\n\nLa app detecto el cambio de estado,\ncongelo la actualizacion del mapa\ny mostro Toast:\n"Se perdio senal GPS"\n\nSin crash. App estable.\n\nResultado: APROBADA', 7, 1.6, 5.5, 4.5, 12, LGRAY)

# ===== SLIDE 7: UI + Videos =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s)
add_text(s, 'Validacion UI/UX — Cap. 20 WebApp', 0.5, 0.2, 12, 0.6, 28, WHITE, True)
add_card(s, 0.4, 1, 6, 5.8)
add_text(s, 'CP-UI-01: Tokens de Diseno Atomico', 0.6, 1.1, 5.5, 0.4, 14, BLUE2, True)
add_text(s, 'Tokens verificados:\n- --color-primary-700: #1a3a6b\n- --color-accent-green: #34c759\n- --font-family-base: Inter\n- --radius-md: 16px\n\nComponentes aislados:\nToast.js / Navbar.js / Icons.js\n\nToast <= 85% ancho pantalla\nResponsive max-width: 480px\n\nResultado: APROBADA', 0.6, 1.6, 5.5, 4.5, 12, LGRAY)
add_card(s, 6.8, 1, 6, 5.8)
add_text(s, 'Videos de Evidencia', 7, 1.1, 5.5, 0.4, 14, GREEN, True)
add_text(s, 'Cada prueba fue grabada en video:\n\n1. CP-AUTH-01_Login_Exitoso.mp4\n2. CP-AUTH-02_Prevencion_XSS.mp4\n3. CP-PERF-01_Foto_Perfil.mp4\n4. CP-COMPRA-01_Saldo_Exacto.mp4\n5. CP-COMPRA-02_Saldo_Insuficiente.mp4\n6. CP-ROL-01_Restriccion_Admin.mp4\n7. CP-ROL-02_Directorio_Usuarios.mp4\n8. CP-GEO-01_GPS_Tracking.mp4\n9. CP-GEO-02_Corte_GPS.mp4\n10. CP-UI-01_Tokens_Diseno.mp4', 7, 1.6, 5.5, 4.5, 12, LGRAY)

# ===== SLIDE 8: Conclusiones =====
s = prs.slides.add_slide(prs.slide_layouts[6])
set_bg(s, BLUE)
add_text(s, 'Conclusiones', 0.5, 0.3, 12.3, 0.6, 30, WHITE, True, PP_ALIGN.CENTER)
add_card(s, 0.8, 1.2, 11.7, 5.2)
add_text(s, '- 10 de 10 pruebas aprobadas (100%)\n- Cero incidentes de Severidad 1\n- Cobertura completa: Auth, Compras, Roles, Perfil, GPS, UI\n\n- Pruebas Criticas (AUTH, COMPRA, ROL): 100%\n- Sin caidas, fugas de datos ni corrupcion de saldos\n- Codigo cumple SOLID, Early Return y SoC\n- Tokens de diseno atomico validados\n\nEl sistema PremiumBus cumple los criterios de aceptacion\nde Pressman y es apto para produccion.', 1.2, 1.6, 10.8, 4.4, 15, LGRAY)
add_text(s, 'PremiumBus — Mayo 2026', 0.5, 6.8, 12.3, 0.3, 12, GRAY, False, PP_ALIGN.CENTER)

# Guardar
out = os.path.join(os.path.dirname(__file__), 'Presentacion_Pruebas_PremiumBus.pptx')
prs.save(out)
print(f'Presentacion guardada en: {out}')

File: Entrega Sistema USB/02_Codigo_Fuente/pruebas/generar_pptx.js
const PptxGenJS = require('pptxgenjs');
const path = require('path');

const pptx = new PptxGenJS();
pptx.layout = 'LAYOUT_WIDE';

const D='0F172A', B='1A3A6B', C='1E293B', W='FFFFFF', L='E2E8F0';
const GR='94A3B8', B2='2B5EA7', GN='22C55E', RD='EF4444', Y='F59E0B';

function prueba(id, titulo, tipo, descripcion, pasos, resultado, colorTit) {
  var s = pptx.addSlide();
  s.background = {color: D};

  // Barra superior con ID
  s.addText(id, {x:0.4, y:0.2, w:2, h:0.45, fontSize:16, color:W, bold:true, fill:{color:B}});
  s.addText(tipo, {x:2.6, y:0.2, w:3, h:0.45, fontSize:11, color:GR, valign:'middle'});
  // Badge resultado
  s.addText('APROBADA', {x:10.5, y:0.2, w:2.3, h:0.45, fontSize:13, color:W, bold:true, fill:{color:GN}, align:'center'});

  // Titulo
  s.addText(titulo, {x:0.4, y:0.85, w:12.5, h:0.5, fontSize:22, color:colorTit || B2, bold:true});

  // --- COLUMNA IZQUIERDA: Info ---
  // Descripcion
  s.addText('Descripcion', {x:0.4, y:1.6, w:5.6, h:0.35, fontSize:11, color:Y, bold:true});
  s.addText(descripcion, {x:0.4, y:2, w:5.6, h:1.8, fontSize:11, color:L, fill:{color:C}, valign:'top'});

  // Pasos
  s.addText('Procedimiento', {x:0.4, y:4, w:5.6, h:0.35, fontSize:11, color:Y, bold:true});
  s.addText(pasos, {x:0.4, y:4.4, w:5.6, h:2, fontSize:11, color:L, fill:{color:C}, valign:'top'});

  // Resultado
  s.addText('Resultado Esperado', {x:0.4, y:6.5, w:5.6, h:0.35, fontSize:11, color:Y, bold:true});
  s.addText(resultado, {x:0.4, y:6.85, w:5.6, h:0.5, fontSize:11, color:GN, fill:{color:C}});

  // --- COLUMNA DERECHA: Espacio para captura ---
  s.addText('CAPTURA DE PANTALLA', {x:6.4, y:1.6, w:6.4, h:0.35, fontSize:11, color:GR, bold:true, align:'center'});
  // Marco placeholder
  s.addText('[ Insertar captura aqui ]\n\nClic derecho > Insertar imagen\no arrastrar captura a este espacio', {
    x:6.4, y:2, w:6.4, h:5.35,
    fontSize:14, color:'475569',
    align:'center', valign:'middle',
    fill:{color:'141E30'},
    border:{type:'dash', pt:2, color:'334155'}
  });
}

// ===== SLIDE 1: Portada =====
var s = pptx.addSlide();
s.background = {color: B};
s.addText('Plan de Pruebas de Software', {x:0.5, y:1.2, w:12.3, h:1, fontSize:40, color:W, bold:true, align:'center'});
s.addText('PremiumBus', {x:0.5, y:2.4, w:12.3, h:0.6, fontSize:24, color:L, align:'center'});
s.addText('Metodologia Pressman (Cap. 17-20)', {x:0.5, y:3.2, w:12.3, h:0.5, fontSize:16, color:GR, align:'center'});
s.addText('10 pruebas ejecutadas  |  10 aprobadas  |  100% cobertura', {x:0.5, y:4.5, w:12.3, h:0.4, fontSize:14, color:GN, align:'center'});
s.addText('Mayo 2026', {x:0.5, y:5.8, w:12.3, h:0.3, fontSize:12, color:GR, align:'center'});

// ===== SLIDE 2: CP-AUTH-01 =====
prueba(
  'CP-AUTH-01',
  'Login Exitoso y Retencion de Token',
  'Autenticacion | Caja Negra | Cap.17',
  'Se verifica que el sistema permite iniciar sesion con credenciales validas y que el token de autenticacion se mantiene persistente tras cerrar y reabrir la aplicacion, evitando que el usuario tenga que re-autenticarse.',
  '1. Ingresar credenciales correctas (email + password)\n2. Verificar acceso exitoso al Home\n3. Cerrar la app desde el administrador de tareas\n4. Reabrir la aplicacion',
  'La sesion se reanuda sin pedir credenciales. Token persistente en localStorage.',
  B2
);

// ===== SLIDE 3: CP-AUTH-02 =====
prueba(
  'CP-AUTH-02',
  'Prevencion de Inyeccion XSS / SQL',
  'Seguridad | Caja Negra | Cap.18',
  'Se intenta inyectar codigo malicioso en el campo de email del login para verificar que el sistema bloquea ataques XSS, SQL Injection y HTML Injection antes de realizar cualquier peticion de red.',
  '1. Ingresar como email: <script>alert(1)</script>\n2. Ingresar: \' OR 1=1 --\n3. Ingresar: <img onerror=alert(1)>\n4. Pulsar boton login en cada caso',
  'Bloqueo inmediato local. Mensaje: "Formato de email invalido" (Early Return).',
  B2
);

// ===== SLIDE 4: CP-COMPRA-01 =====
prueba(
  'CP-COMPRA-01',
  'Compra con Saldo Exacto $13.50',
  'Compras | Valores Limite | Cap.18',
  'Se prueba el caso limite donde el saldo del usuario es exactamente igual al precio del boleto ($13.50 MXN). Se verifica que la transaccion se aprueba y el saldo queda en $0.00 sin errores de punto flotante.',
  '1. Configurar saldo del usuario = $13.50\n2. Ir a seccion de tickets\n3. Seleccionar ruta y asiento\n4. Pulsar boton "Adquirir"',
  'Transaccion aprobada. Codigo QR emitido. Saldo final: $0.00 exacto.',
  GN
);

// ===== SLIDE 5: CP-COMPRA-02 =====
prueba(
  'CP-COMPRA-02',
  'Rechazo por Saldo Insuficiente $13.49',
  'Compras | Valores Limite | Cap.18',
  'Se prueba el caso donde el saldo es $13.49 (1 centavo por debajo del precio). El sistema debe rechazar la compra inmediatamente mostrando un Toast de error, sin realizar deduccion alguna.',
  '1. Configurar saldo del usuario = $13.49\n2. Intentar comprar boleto de $13.50\n3. Observar respuesta del sistema',
  'Toast rojo: "Fondo insuficiente". Saldo preservado sin deduccion.',
  RD
);

// ===== SLIDE 6: CP-ROL-01 =====
prueba(
  'CP-ROL-01',
  'Restriccion de Vistas para Administrador',
  'Roles | Orientado a Objetos | Cap.19',
  'Se verifica que al iniciar sesion como Administrador, el boton "Comprar Boletos" desaparece de la interfaz mediante renderizado condicional. Los administradores no deben poder comprar.',
  '1. Iniciar sesion con cuenta tipo Admin\n2. Inspeccionar pantalla de inicio\n3. Verificar menu lateral\n4. Buscar opcion "Comprar"',
  'Boton "Comprar" OCULTO. Panel Conductores y Estadisticas VISIBLES.',
  B2
);

// ===== SLIDE 7: CP-ROL-02 =====
prueba(
  'CP-ROL-02',
  'Acceso al Directorio de Usuarios',
  'Roles | Orientado a Objetos | Cap.19',
  'Se verifica que un Administrador puede acceder al directorio completo de usuarios registrados. La consulta debe estar optimizada sin requerir indices compuestos en Firebase.',
  '1. Iniciar sesion como Admin\n2. Navegar a "Directorio de Usuarios"\n3. Verificar que la lista se carga completa\n4. Revisar consola por errores de indices',
  'Lista completa desplegada. Sin errores de composite indices en Firebase.',
  B2
);

// ===== SLIDE 8: CP-PERF-01 =====
prueba(
  'CP-PERF-01',
  'Carga Masiva de Fotografia de Perfil',
  'Rendimiento | Firebase Storage | Cap.17',
  'Se verifica la subida asincrona de una imagen JPG al perfil del usuario. Debe mostrar barra de progreso durante la carga y actualizar la miniatura del avatar en tiempo real al completar.',
  '1. Clic en "Modificar Perfil"\n2. Seleccionar imagen JPG valida (< 2MB)\n3. Observar barra de progreso\n4. Verificar actualizacion de miniatura',
  'Barra de progreso fluida. Avatar actualizado en tiempo real sin recargar.',
  B2
);

// ===== SLIDE 9: CP-GEO-01 =====
prueba(
  'CP-GEO-01',
  'Seguimiento GPS en Tiempo Real',
  'Geolocalizacion | WebApp/Movil | Cap.20',
  'Se verifica que el mapa muestra el marcador del autobus desplazandose fluidamente sin saltos exagerados. Las coordenadas y velocidad deben actualizarse en tiempo real mediante un poller optimizado.',
  '1. Activar GPS y dar permisos\n2. Abrir seccion de Mapa\n3. Observar movimiento del marcador del bus\n4. Verificar actualizacion de coordenadas',
  'Marcador se desplaza fluidamente. Coordenadas actualizadas sin saltos.',
  B2
);

// ===== SLIDE 10: CP-GEO-02 =====
prueba(
  'CP-GEO-02',
  'Corte Abrupto de Servicios GPS',
  'Geolocalizacion | Resiliencia | Cap.20',
  'Se desactiva el GPS mientras el tracking esta activo. La app debe detectar la perdida de senal, congelar la actualizacion del mapa y notificar al usuario via Toast sin provocar un crash.',
  '1. Tener el mapa funcionando normalmente\n2. Ir a Ajustes del sistema y desactivar GPS\n3. Regresar a la app\n4. Observar comportamiento',
  'Toast: "Se perdio senal GPS". Mapa congelado. Sin crash.',
  RD
);

// ===== SLIDE 11: CP-UI-01 =====
prueba(
  'CP-UI-01',
  'Validacion de Tokens de Diseno Atomico',
  'UI/UX | WebApp | Cap.20',
  'Se verifica que toda la aplicacion cumple con los tokens de diseno definidos: colores, tipografia (Inter), border-radius y componentes aislados. Los Toast no deben superar el 85% del ancho de pantalla.',
  '1. Inspeccionar variables CSS del sistema\n2. Verificar uso de tokens en componentes\n3. Medir ancho de Toast en pantalla pequena\n4. Validar responsive max-width: 480px',
  'Todos los tokens aplicados. Toast <= 85% ancho. Componentes aislados.',
  B2
);

// ===== SLIDE 12: Conclusiones =====
s = pptx.addSlide();
s.background = {color: B};
s.addText('Conclusiones', {x:0.5, y:0.3, w:12.3, h:0.7, fontSize:30, color:W, bold:true, align:'center'});
s.addText(
  '- 10 de 10 pruebas aprobadas (100%)\n' +
  '- Cero incidentes de Severidad 1\n' +
  '- Cobertura completa: Auth, Compras, Roles, Perfil, GPS, UI\n\n' +
  '- Pruebas Criticas (AUTH, COMPRA, ROL): 100%\n' +
  '- Sin caidas, fugas de datos ni corrupcion de saldos\n' +
  '- Codigo cumple SOLID, Early Return y SoC\n' +
  '- Tokens de diseno atomico validados\n\n' +
  'El sistema PremiumBus cumple los criterios de aceptacion\n' +
  'de Pressman y es apto para produccion.',
  {x:1, y:1.3, w:11.3, h:5, fontSize:16, color:L, fill:{color:C}, valign:'top'}
);
s.addText('PremiumBus - Mayo 2026', {x:0.5, y:6.8, w:12.3, h:0.3, fontSize:12, color:GR, align:'center'});

// GUARDAR
var outPath = path.join(__dirname, 'Presentacion_Pruebas_PremiumBus.pptx');
pptx.writeFile({fileName: outPath}).then(function(){
  console.log('Presentacion guardada en: ' + outPath);
}).catch(function(e){
  console.error('Error:', e);
});
